src/exec/execute_pipeline.c

```diff
@@ -18,25 +18,56 @@
 #include "exec_internal.h"
 #include "utils.h"
 
-#define PIPE_IN 0
-#define PIPE_OUT 1
+#define PIPE_READ 0
+#define PIPE_WRITE 1
 
-pid_t	execute_pipeline(t_ast *node, t_env_list *env_list,
-			int fd_in, int fd_out)
+int	execute_left_node(t_ast *node, t_env_list *env_list,
+		t_pipeline_conf *conf, int *pipe_fd)
 {
-	int				pipe_fds[2];
-	pid_t			pid;
+	int	ret;
 
-	if (node->type == AST_COMMAND)
-		return (execute_simple_command(node, env_list, fd_in, fd_out));
-	if (pipe(pipe_fds) == -1)
+	conf->fd_out = pipe_fd[PIPE_WRITE];
+	ret = execute_ast_node(node, env_list, conf);
+	close(pipe_fd[PIPE_WRITE]);
+	return (ret);
+}
+
+int	execute_right_node(t_ast *node, t_env_list *env_list,
+		t_pipeline_conf *conf, int *pipe_fd)
+{
+	int	ret;
+
+	if (conf->fd_in != STDIN_FILENO)
+		close(conf->fd_in);
+	conf->fd_in = pipe_fd[PIPE_READ];
+	if (node->type != AST_PIPE)
+		conf->fd_out = STDOUT_FILENO;
+	ret = execute_ast_node(node, env_list, conf);
+	close(pipe_fd[PIPE_READ]);
```

Comment:
redundant `close()`, since Line40,41 has done it already.
```suggestion
```

Reply:
It is necessary.
Usually `pipe_fd[PIPE_READ]` will be closed at line 41
but `pipe_fd[PIPE_READ]` of last command in the pipeline won't be closed at the line.
So line 46 is necessary.
@alex216

Reply:
## Experiment

### Current implementation

``` bash
minishell$ echo Hello | cat
Hello
minishell$ echo Hello | cat
Hello
minishell$ echo Hello | cat
Hello
minishell$ echo Hello | cat
Hello
minishell$ echo Hello | cat
Hello

# In another terminal

❯ lsof -c minishell
COMMAND     PID       USER   FD   TYPE DEVICE SIZE/OFF     NODE NAME
minishell 59046 reiyaasuke  cwd    DIR   1,14      800 68945239 /Users/reiyaasuke/clang/minishell
minishell 59046 reiyaasuke  txt    REG   1,14   383704 75835580 /Users/reiyaasuke/clang/minishell/minishell
minishell 59046 reiyaasuke    0u   CHR  16,37 0t160534     6961 /dev/ttys037
minishell 59046 reiyaasuke    1u   CHR  16,37 0t160534     6961 /dev/ttys037
minishell 59046 reiyaasuke    2u   CHR  16,37 0t160534     6961 /dev/ttys037
```

There are no fd leaks.

### Proposed implementation (delete line 46)

``` bash
minishell$ echo Hello | cat
Hello
minishell$ echo Hello | cat
Hello
minishell$ echo Hello | cat
Hello
minishell$ echo Hello | cat
Hello
minishell$ echo Hello | cat
Hello

# In another terminal
❯ lsof -c minishell
COMMAND     PID       USER   FD   TYPE             DEVICE SIZE/OFF     NODE NAME
minishell 59572 reiyaasuke  cwd    DIR               1,14      800 68945239 /Users/reiyaasuke/clang/minishell
minishell 59572 reiyaasuke  txt    REG               1,14   383704 75837356 /Users/reiyaasuke/clang/minishell/minishell
minishell 59572 reiyaasuke    0u   CHR              16,37 0t171299     6961 /dev/ttys037
minishell 59572 reiyaasuke    1u   CHR              16,37 0t171299     6961 /dev/ttys037
minishell 59572 reiyaasuke    2u   CHR              16,37 0t171299     6961 /dev/ttys037
minishell 59572 reiyaasuke    3   PIPE 0xa093add1cdd27c2d    16384
minishell 59572 reiyaasuke    4   PIPE 0x7d67927a3a8532cf    16384
minishell 59572 reiyaasuke    5   PIPE 0x611672c7b0505e43    16384
minishell 59572 reiyaasuke    6   PIPE 0xe2ade9d1fd4af972    16384
```

There are some fd leaks.